content/posts/dns-command/index.md

```diff
@@ -0,0 +1,113 @@
+With networksetup, changing my DNS config is as simple as
+
+```bash
+# Adding 1.1.1.1 to my nameservers
+networksetup -setdnsservers Wi-Fi 1.1.1.1
+# Clear all my alternate nameservers
+networksetup -setdnsservers Wi-Fi empty
+```
+
+Now I can easily configure my DNS resolvers to be the same as the instructions for 1.1.1.1 by
+
+```bash
+# 4 nameservers - 2 IPv4 & 2 IPv6
+networksetup -setdnsservers Wi-Fi 1.1.1.1 1.0.0.1 2606:4700:4700::1111 2606:4700:4700::1001
+```
+
+But this is quite lengthy to type as well, even though I would prefer this over Network Preferences UI any day. So I went an another level of laziness by wrapping it inside a bash function
+
+```bash
+function dns1111() {
+    networksetup -setdnsservers Wi-Fi 1.1.1.1 1.0.0.1 2606:4700:4700::1111 2606:4700:4700::1001
+}
+```
+
+This seems to be pretty neat and does the job, but hold on a second. How about switching it back to empty?
+A flag/argument to this bash function would do the job fine but I want to use the least effort as possible and save myself the mental burden of remembering whether my alternate DNS config is empty.
+So I thought to myself: "Why not detect the current state and switch the DNS config based on that? Like a literal switch button - you **don't** have to tell what state you want it to be, just the opposite of what comes before".
+With that in mind, I found this option for networksetup that would tell me the current state
+
+```bash
+networksetup -getdnsservers Wi-Fi
+```
+
+It would list all of my alternate DNS servers, or print out this string `There aren't any DNS Servers set on Wi-Fi` if there isn't any.
+And now I can complete my function, with some cool emojis to signify whether the config has just been applied or removed
+
+```bash
+function dns1111() {
+  if [[ $(networksetup -getdnsservers Wi-Fi) == "There aren't any DNS Servers set on Wi-Fi"* ]]; then
+    networksetup -setdnsservers Wi-Fi 1.1.1.1 1.0.0.1 2606:4700:4700::1111 2606:4700:4700::1001 && echo 🚀 🚀 🚀
+  else
+    networksetup -setdnsservers Wi-Fi empty && echo 🚦 🚦 🚦
+  fi
+}
+```
+
+Now I can easily switch back and forth between 1.1.1.1 and default DNS configs with just a short command `dns1111`
+
+### Bonus: Linux based machine
+
+Since `networksetup` is not available not Linux based OSes, we would need to modify `/etc/resolv.conf` directly to add & remove the alternate DNS.
+I have included my implementation below.
+If you have a better approach, don't hesitate to reach out to me at [hung.ngn.the@gmail.com](mailto:hung.ngn.the@gmail.com)
+
+```bash
+function dns1111() {
+  # detect existence of the backup file to see whether the config is in place
+  if [ -e /etc/resolv.conf.bak ]]; then
+    mv /etc/resolv.conf.bak /etc/resolv.conf && echo 🚦 🚦 🚦
+  else
+    networksetup -setdnsservers Wi-Fi empty
+    echo "nameserver 1.1.1.1
+nameserver 1.0.0.1
+nameserver 2606:4700:4700::1111
+nameserver 2606:4700:4700::1001" >> /etc/resolv.conf
+    && echo 🚀 🚀 🚀
+  fi
+}
+```
```

Comment:
Network managers tend to overwrite /etc/resolv.conf in some distro. To prevent programs from overwriting /etc/resolv.conf, it is also possible to write-protect it by setting the immutable file attribute:
```bash
# chattr +i /etc/resolv.conf
```
TIPS:  If you want multiple processes to write to /etc/resolv.conf, you can use `resolvconf`.

Reply:
There seems to be an even better alternative https://unix.stackexchange.com/questions/128220/how-do-i-set-my-dns-when-resolv-conf-is-being-overwritten/163506#163506
What do u think? 🤔